sys/kern/vm_map.c

```diff
@@ -493,21 +531,29 @@ int vm_page_fault(vm_map_t *map, vaddr_t fault_addr, vm_prot_t fault_type) {
 
   assert(ent->start <= fault_addr && fault_addr < ent->end);
 
-  vm_object_t *obj = ent->object;
+  vaddr_t fault_page = fault_addr & -PAGESIZE;
 
-  assert(obj != NULL);
+  if (!ent->aref.amap) {
+    int slots = vaddr_to_slot(ent->end - ent->start);
+    ent->aref.offset = 0;
+    ent->aref.amap = vm_amap_alloc(slots);
+  }
 
-  vaddr_t fault_page = fault_addr & -PAGESIZE;
-  vaddr_t offset = ent->offset + (fault_page - ent->start);
-  vm_page_t *frame = vm_object_find_page(ent->object, offset);
+  bool insert = false;
+  int offset = vaddr_to_slot(fault_page - ent->start);
+  vm_page_t *frame = vm_amap_find_page(ent->aref, offset);
 
-  if (frame == NULL)
-    frame = obj->vo_pager->pgr_fault(obj, offset);
+  if (frame == NULL) {
+    insert = true;
+    frame = vm_page_alloc_zero(1);
```

Comment:
If that's happen to be the only reference to `vm_page_alloc_zero` then please remove the function and zero out the page here.